funsor/expectation.py

```diff
@@ -0,0 +1,97 @@
+from __future__ import absolute_import, division, print_function
+
+from collections import OrderedDict, defaultdict
+
+import funsor.ops as ops
+from funsor.domains import reals
+from funsor.interpreter import interpretation, reinterpret
+from funsor.optimizer import associate, Finitary
+from funsor.terms import Funsor, Unary, eager, monte_carlo, to_funsor, Reduce
+
+
+class Expectation(Funsor):
+    """
+    Expectation of an ``integrand`` with resepct to a nonnegative ``measure``.
+    """
+    def __init__(self, measure, integrand, reduced_vars):
+        assert isinstance(measure, Funsor)
+        assert isinstance(integrand, Funsor)
+        assert isinstance(reduced_vars, Funsor)
+        assert measure.output == reals()
+        inputs = OrderedDict((k, d) for part in (measure, integrand)
+                             for k, d in part.inputs.items()
+                             if k not in reduced_vars)
+        output = integrand.output
+        super(Expectation, self).__init__(inputs, output)
+        self.measure = measure
+        self.integrand = integrand
+        self.reduced_vars = reduced_vars
+
+    def eager_subs(self, subs):
+        raise NotImplementedError('TODO')
+
+
+@eager.register(Expectation, Funsor, Funsor, frozenset)
+def eager_expectation(measure, integrand, reduced_vars):
+    return (measure * integrand).reduce(ops.add, reduced_vars)
+
+
+@monte_carlo.register(Expectation, Funsor, Funsor, frozenset)
```

Comment:
@eb8680 @martinjankowiak Here is a sketch of a monte carlo `Expectation` implementation that matches probability factors to the terms they depend on, as in `TraceGraph_ELBO` or `Dice`. This approach isn't quite right, but it is a step towards a feasible design. Thoughts?

Reply:
I spent this afternoon working through implementing a similar `Integrate` class to help myself understand all the issues involved.  I think something like this is basically what we want - let's talk tomorrow.